domain/permission/state/state.go

```diff
@@ -0,0 +1,385 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"database/sql"
+	"time"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+	"github.com/juju/names/v5"
+
+	coredatabase "github.com/juju/juju/core/database"
+	corepermission "github.com/juju/juju/core/permission"
+	"github.com/juju/juju/domain"
+	"github.com/juju/juju/domain/permission"
+	permissionerrors "github.com/juju/juju/domain/permission/errors"
+	usererrors "github.com/juju/juju/domain/user/errors"
+	databaseutils "github.com/juju/juju/internal/database"
+	"github.com/juju/juju/internal/uuid"
+)
+
+// Logger is the interface used by the state to log messages.
+type Logger interface {
+	Debugf(string, ...interface{})
+}
+
+// State describes retrieval and persistence methods for storage.
+type State struct {
+	*domain.StateBase
+	logger Logger
+}
+
+// NewState returns a new state reference.
+func NewState(factory coredatabase.TxnRunnerFactory, logger Logger) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+		logger:    logger,
+	}
+}
+
+// User represents a user in the system where the values overlap with
+// corepermission.UserAccess.
+type User struct {
+	// UUID is the unique identifier for the user.
+	UUID string `db:"uuid"`
+
+	// Name is the username of the user.
+	Name string `db:"name"`
+
+	// DisplayName is a user-friendly name represent the user as.
+	DisplayName string `db:"display_name"`
+
+	// CreatorName is the name of the user that created this user.
+	CreatorName string `db:"created_by_name"`
+
+	// CreatedAt is the time that the user was created at.
+	CreatedAt time.Time `db:"created_at"`
+
+	// Disabled is true if the user is disabled.
+	Disabled bool `db:"disabled"`
+}
+
+// toCoreUserAccess converts the state user to a core permission UserAccess.
+func (u User) toCoreUserAccess() corepermission.UserAccess {
+	return corepermission.UserAccess{
```

Comment:
Add the permission UUID to this type and return the UUID you created with it. We won't use it now, but potentially in the future.